### PR TITLE
Reporter no longer runs SimpleCov, but aggregates/posts results instead

### PR DESCRIPTION
**Pull request by mrb**
Due to a history of legacy functionality changes with the reporter and
on codeclimate.com, this reporter had one major flaw that because of
the circumstances didn't really bother anyone: if tests failed or errors
occurred during a test run, the reporter would still report coverage
to codeclimate.com. This has started to become a problem since we have
released a browser extension and other features which leverage test reports
on non-master or 'default' branches.

The solution is to make this reporter similar to the others for the major
languages we support. Instead of including a snippet in a test/spec helper
which initiates and runs SimpleCov and then posts results to codeclimate,
users will now do what they did before: include the SimpleCov snippet in
their helpers. After tests, users will execute the added bin/codeclimate-ruby
binary, which will aggregate and format results before posting them to
codeclimate. This obviates the "partial results" problem because your CI
script will not reach the line where you run bin/codeclimate-ruby if there
is an error or failure.

We will now have parity between how this reporter is used and how others
are used, and the presenting issues should be mitigated.

### CODE REVIEW

**Comment by wfleming** [comment]
Should README changes be addressed here or is that coming later?

Code LGTM.

File: lib/code_climate/test_reporter.rb [hunk 1]
```diff
@@ -1,16 +1,5 @@
 module CodeClimate
   module TestReporter
-    def self.start
```
Comment (wfleming) [inline lib/code_climate/test_reporter.rb]: Maybe we should keep this around but have it output a deprecated log message & exit? I'm sure there will be users who upgrade the gem in their gemfile at some point & wonder why they're not getting results.

**Comment by mrb** [comment]
@wfleming Good points both re: README and logging the deprecated method. Will add those.

**Comment by mrb** [comment]
@codeclimate/review Updated with @wfleming's suggestion for printing a deprecation warning and cleaning up the README. Decided to remove all of the specific information in the README in lieu of keeping one up to date document on our docs site, and linking to that.

File: bin/codeclimate-ruby [hunk 2]
```diff
@@ -0,0 +1,19 @@
+#!/usr/bin/env ruby
```
Comment (pbrisbin) [inline bin/codeclimate-ruby]: I think we call this `codeclimate-test-reporter` in the other reporters.

File: bin/codeclimate-ruby [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+    result = JSON.parse(File.read(COVERAGE_FILE))
+
+    CodeClimate::TestReporter::Formatter.new.format(result)
+  else
+    $stderr.puts "Coverage results not found"
+    exit(1)
+  end
+else
+  $stderr.puts "Cannot post results: environment variable CODECLIMATE_REPO_TOKEN must be set."
+  exit(1)
```
Comment (pbrisbin) [inline bin/codeclimate-ruby]: I might consider this a warn-but-still-exit-0 case, WDYT?

The other case, where you have the env var set (so you intend to submit) but we don't find a file, make sense to me as a non-zero exit.

File: bin/codeclimate-ruby [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+
+require "codeclimate-test-reporter"
+
+COVERAGE_FILE = "coverage/.resultset.json".freeze
+
+if ENV["CODECLIMATE_REPO_TOKEN"]
+  if File.exist?(COVERAGE_FILE)
+    result = JSON.parse(File.read(COVERAGE_FILE))
+
+    CodeClimate::TestReporter::Formatter.new.format(result)
```
Comment (pbrisbin) [inline bin/codeclimate-ruby]: Should you do some rescues here to make the failure cases (JSON parse-error or the reporter blowing up) more readable than a bare backtrace?

File: lib/code_climate/test_reporter.rb [hunk 5]
```diff
@@ -1,14 +1,17 @@
...
     def self.start
-      if run?
-        require "simplecov"
-        ::SimpleCov.add_filter "vendor"
-        ::SimpleCov.formatter = Formatter
-        ::SimpleCov.start(configuration.profile) do
-          skip_token CodeClimate::TestReporter.configuration.skip_token
-        end
-      end
+      logger.warn(WARNING_MESSAGE)
```
Comment (pbrisbin) [inline lib/code_climate/test_reporter.rb]: Nice touch!

File: lib/code_climate/test_reporter/formatter.rb [hunk 6]
```diff
@@ -15,9 +15,11 @@ class Formatter
       def format(result)
         return true unless CodeClimate::TestReporter.run?
 
-        print "Coverage = #{result.source_files.covered_percent.round(2)}%. "
+        simplecov_result = SimpleCov::Result.new(result.first[1]["coverage"])
```
Comment (pbrisbin) [inline lib/code_climate/test_reporter/formatter.rb]: Ideally I think we should have a nice error for an input file we can't handle. I think I would `raise` our own error here if there's no `1` index, no `"coverage"` key, or `SimpleCov::Result.new` itself raises. WDYT?

**Comment by pbrisbin** [comment]
Some minor questions about the executable name and general error-handling approach. LGTM me besides.

**Comment by mrb** [comment]
@pbrisbin I considered better error handling but wanted to see if it bothered anyone else that it wasn't better. Since it seems to bug you too I'll work on that.

File: lib/code_climate/test_reporter/formatter.rb (hunk 6 above)
Comment (pbrisbin) [inline lib/code_climate/test_reporter/formatter.rb]: Hmm. This is a fun little bikeshed. I might use exception handling:

```rb
  result.fetch(0).fetch(1).fetch("coverage")
rescue NoMethodError, KeyError => ex
  raise InvalidSimpleCovResultError, ex.message
end
```

WDYT of something along those lines?

Comment (mrb) [inline lib/code_climate/test_reporter/formatter.rb]: @pbrisbin Yeah that looks good, I kinda forgot fetch can work like that.

**Comment by mrb** [comment]
@pbrisbin Ping from here!

File: bin/codeclimate-test-reporter [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+require "codeclimate-test-reporter"
+
+COVERAGE_FILE = "coverage/.resultset.json".freeze
+
+if ENV["CODECLIMATE_REPO_TOKEN"]
+  if File.exist?(COVERAGE_FILE)
+    begin
+      results = JSON.parse(File.read(COVERAGE_FILE))
+    rescue JSON::ParserError => e
+      $stderr.puts "Error encountered while parseing #{COVERAGE_FILE}: #{e}"
```
Comment (pbrisbin) [inline bin/codeclimate-test-reporter]: Typo: parsing

File: lib/code_climate/test_reporter/shorten_filename.rb [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+        return @filename unless ::SimpleCov.root
+        apply_prefix @filename.gsub(/^#{::SimpleCov.root}/, ".").gsub(%r{^\./}, "")
+      end
+
+      private
+
+      def apply_prefix(filename)
+        prefix = CodeClimate::TestReporter.configuration.path_prefix
+        return filename if prefix.nil?
+        "#{prefix}/#{filename}"
```
Comment (pbrisbin) [inline lib/code_climate/test_reporter/shorten_filename.rb]: This feels a bit dense, WDYT of:

```rb
if (prefix = ...)
  File.join(prefix, filename)
else
  filename
end
```

**Comment by mrb** [comment]
@pbrisbin I addressed your last comments and squashed the commits.

**Comment by gerrywastaken** [comment]
It seems others didn't receive the memo: https://docs.travis-ci.com/user/code-climate/

It was tricky to figure out why my builds suddenly started failing.

**Comment by maxjacobson** [comment]
@gerrywastaken thanks for the head's up about that. I'll reach out to Travis CI.

**Pull request merged**